components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +87,166 @@ public AbstractHTTPFunction() {
+                asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+            } catch (Exception e) {
+                LOG.error("Error while applying authentication to the request.", e);
+                asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+            }
+        });
+        JsGraphBuilder.addLongWaitProcess(asyncProcess, eventHandlers);
+    }
+
+    /**
+     * Execute the request with retries.
+     *
+     * @param request     HttpUriRequest.
+     * @param endpointURL Endpoint URL.
+     * @param maxRetries  Maximum number of retries.
+     * @return Pair of outcome and json.
+     */
+    private Pair<String, JSONObject> executeRequestWithRetries(HttpUriRequest request, String endpointURL, int maxRetries) {
+
+        JSONObject json = null;
+        String outcome = Constants.OUTCOME_FAIL;
+        int attempts = 0;
+
+        while (attempts < maxRetries) {
+            attempts++;
+            LOG.warn("Retrying the request. Attempt: " + attempts + " for endpoint: " + endpointURL);
+            if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                        DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                        Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.EXTERNAL_API, endpointURL)
+                        .resultMessage("Retrying the request for external endpoint.")
+                        .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                        .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+            }
+            Pair<String, JSONObject> result = executeRequest(request, endpointURL);
+            if (result.getLeft().equals(Constants.OUTCOME_SUCCESS)) {
+                if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                    DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                            DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                            Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                    diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.EXTERNAL_API, endpointURL)
+                            .resultMessage("Successfully called the external endpoint.")
+                            .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                            .resultStatus(DiagnosticLog.ResultStatus.SUCCESS);
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
                 }
+                return result;
+            }
+        }
+        return Pair.of(outcome, json); // Return outcome and json (which might be null if never successful)
+    }
```

Comment:
I think you can still improve this. And the context of whether it's a response error or a request time out should be identified in order to add that context to the diagnostic log

Reply:
Fixed with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/05c4f77e8d18ba911c3b98bc7e18ec3fb73dcd33